pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,274 @@
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = "0"
+	}
+
+	attemptsMade, convErr := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+	if convErr != nil {
+		// We shouldn't get an error here, but if we do, we will set verification attempts to max allowed
+		// so that we at least now have a valid value, and let the workflow continue to the
+		// subsequent attempts check
+		log.Error(ctx, convErr, fmt.Sprintf("error converting annotation [%s] value [%s] to integer, on UserSignup: [%s]",
+			v1alpha1.UserVerificationAttemptsAnnotationKey,
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey], signup.Name))
+		attemptsMade = s.config.GetVerificationAttemptsAllowed()
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+	}
+
+	// If the user has made more attempts than is allowed per day, return an error
+	if attemptsMade >= s.config.GetVerificationAttemptsAllowed() {
+		verificationErr = errors.NewTooManyRequestsError("too many verification attempts", "")
+	}
+
+	if verificationErr == nil {
+		// Parse the verification expiry timestamp
+		exp, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+		if parseErr != nil {
+			// If the verification expiry timestamp is corrupt or missing, then return an error
+			verificationErr = errors.NewInternalError(parseErr, "error parsing expiry timestamp")
+		} else if now.After(exp) {
+			// If it is now past the expiry timestamp for the verification code, return a 403 Forbidden error
+			verificationErr = errors.NewForbiddenError("expired", "verification code expired")
+		}
+	}
+
+	if verificationErr == nil {
+		if code != signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
+			// The code doesn't match
+			attemptsMade++
+			signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+			verificationErr = errors.NewForbiddenError("invalid code", "the provided code is invalid")
+		}
+	}
+
+	if verificationErr == nil {
+		// If the code matches then set VerificationRequired to false, reset other verification annotations
+		signup.Spec.VerificationRequired = false
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCodeAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationAttemptsAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCounterAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationInitTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationExpiryAnnotationKey)
+	} else {
+		log.Error(ctx, verificationErr, "error validating verification code")
```

Comment:
we log this error here but we never return it - is there a reason for this? later, verificationErr is re-assigned so the fact that this error even occurred will be lost..

Reply:
The error is returned further down, assuming that the `UserSignup` update doesn't return an error itself.

Reply:
oh, i forgot, this is because we want to update the usersignup even if the verification failed. instead of having error messages take precedence, maybe we really should be returning an array of errors? as a user, if the validation failed, i'd want to know that because it's something i can fix on my end. the usersignup update failing isn't something that's in my control though...